### PR TITLE
[cdac] GetMethodDescData for jitted methods

### PR DESCRIPTION
**Pull request by lambdageek**
This is enough for `!PrintException` without R2R methods on the stack

There's also a `ReJIT` contract here which just checks whether rejit is enabled.

Most of the complexity is in validating MethodDescs

Contributes to https://github.com/dotnet/runtime/issues/99302
Contributes to https://github.com/dotnet/runtime/issues/108553

### CODE REVIEW

File: src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Legacy/ICorProfiler.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Marshalling;
+
+namespace Microsoft.Diagnostics.DataContractReader.Legacy;
+
+internal enum COR_PRF_MONITOR
```
Comment (AaronRobinsonMSFT) [inline src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Legacy/ICorProfiler.cs:10]: This file name seems off.

File: src/native/managed/cdacreader/src/Legacy/SOSDacImpl.cs [hunk 2]
```diff
@@ -190,21 +190,158 @@ int ISOSDacInterface.GetMethodDescData(ulong methodDesc, ulong ip, DacpMethodDes
             // elements we return
             return HResults.E_INVALIDARG;
         }
+        int hr = HResults.E_NOTIMPL;
+
+
```
Comment (AaronRobinsonMSFT) [inline src/native/managed/cdacreader/src/Legacy/SOSDacImpl.cs:195]:
```suggestion

        int hr = HResults.E_NOTIMPL;
```

File: src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Legacy/ICorProfiler.cs (hunk 1 above)
Comment (lambdageek) [reply src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Legacy/ICorProfiler.cs:10]: This is meant to follow the same naming convention as the other files in this directory: the file is named after the main COM interface and includes the interface definition plus any structs and enums that woudl normally be in the same .idl file.  Except in this case I didn't need any of the rest of ICorProfiler, so I didn't add it.

I can add it

Comment (lambdageek) [reply src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Legacy/ICorProfiler.cs:10]: On second thought this is ridiculous. I'll just move the enun into ReJIT_1, same as other implementation details flags that go with a contract

File: src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/RuntimeTypeSystem_1.cs [hunk 3]
```diff
@@ -758,4 +823,291 @@ private MethodTable GetOrCreateMethodTable(MethodDesc methodDesc)
...
+            Address = address;
+        }
+
+        public TargetPointer GetAddressOfSlot(uint slotNum)
+        {
+            TargetPointer indirectionPointer = Address + (ulong)(slotNum >> NumPointersPerIndirectionLog2) * (ulong)_target.PointerSize;
+            TargetPointer slotsStart = _target.ReadPointer(indirectionPointer);
+            return slotsStart + (ulong)(slotNum & (NumPointersPerIndirection - 1)) * (ulong)_target.PointerSize;
+        }
+
```
Comment (davidwrighton) [inline src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/RuntimeTypeSystem_1.cs:846]: Nit. Empty lines

File: src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/RuntimeTypeSystemHelpers/MethodValidation.cs [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+                return AdditionalPointersHelper(MethodDescFlags_1.MethodDescFlags.HasNonVtableSlot | MethodDescFlags_1.MethodDescFlags.HasMethodImpl);
+            }
+        }
+
+        internal int AdditionalPointersCount => AdditionalPointersHelper(MethodDescFlags_1.MethodDescFlags.MethodDescAdditionalPointersMask);
+        #endregion Additional Pointers
+
+        internal bool HasStableEntryPoint => HasFlags(MethodDescFlags_1.MethodDescFlags3.HasStableEntryPoint);
+        internal bool HasPrecode => HasFlags(MethodDescFlags_1.MethodDescFlags3.HasPrecode);
+
```
Comment (davidwrighton) [inline src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/RuntimeTypeSystemHelpers/MethodValidation.cs:115]: Empty line

**Review by davidwrighton — COMMENTED** [review]
I've read through this and it seems like you've made good progress towards our goal here. I've a few comments about formatting + a few more general comments.

1. I see a lot of code dedicated to `MethodDesc` validation. Is that really necessary, and does the behavior match our existing DAC? We may run this in an environment where the full heap isn't available, and we might want code to do a best effort job of producing results. Also, are there changes to the `MethodDesc` validation logic here, or just refactoring to a new location? Is the validation logic documented in the data contract markdown?
2. I don't see documentation updates in the markdown yet. I see that some existing contracts look to have their implementation modified, but I don't see the corresponding documentation updates.
3. Would it be possible to build a bulleted list of the TODO items that are not assertions that are needed to make this contract work completely?

**Comment by lambdageek** [comment]
> I see a lot of code dedicated to MethodDesc validation. 

yes

> Is that really necessary, and does the behavior match our existing DAC?

I tried to make it match the existing DAC.

Is it necessary?  I'm not sure.  We do need _some_ validation - I have discovered that the `!U` command in SOS doesn't differentiate between a MethodDesc and an IP address and expects to be able to call `GetMethodDescData` with an IP address as the "method desc" argument and get E_INVALIDARG to tell it "this was not a method desc, it was an IP address"


> We may run this in an environment where the full heap isn't available, and we might want code to do a best effort job of producing results.

Understood

> Also, are there changes to the MethodDesc validation logic here, or just refactoring to a new location?

There is additional validation here.  The version checked in previously was an incomplete snapshot (basically just establishing the internal API for validation).

> Is the validation logic documented in the data contract markdown?

No.  Is it part of the contract? I've been thinking of it as an artifact of the reader

**Comment by davidwrighton** [comment]
> > Is the validation logic documented in the data contract markdown?
> 
> No. Is it part of the contract? I've been thinking of it as an artifact of the reader

I don't know. It's certainly something that the implementation of the contract depends on to some extent, so all of the dependencies of it (such as the fields it accesses, etc) should probably be in the contract, but the actual validation logic is a more interesting question. I would leave it out of the contract spec for now, so I wouldn't worry about it this week, but given that SOS actually depends on this logic doing something useful for correct functioning, its implied that something is required to make our diagnostic stack work. I would file a bug that you didn't write contract docs for this and leave it at that.

**Review by davidwrighton — APPROVED** [review]

**Pull request merged**